WT-3848 enhance post conditions of prepared transactions

bvpvamsikrishna opened the pull request:
WIP : For review of approach. In this approach we return error, if transaction state is prepared, we will add this check in all possible API's which should not be supported after a prepare-transaction call.
To enhance post conditions of prepared transactions to check validity of operation invoked.
Only session level API's are modified.
Cursor level API's are not updated to return error yet.

michaelcahill commented on test/suite/test_prepare02.py:
```diff
@@ -0,0 +1,88 @@
+        # Commit after prepare is permitted
+        self.session.begin_transaction()
+        self.session.prepare_transaction()
+        with self.expectedStderrPattern(''):
```
@bvpvamsikrishna, can you please add some cursor operations in here as well (e.g., `c['key'] = 'value' for some cursor `c` opened in `self.session`)?

michaelcahill replied:
Sorry, I see in the pull request description that cursor operations are not yet checked.

It seems like that wouldn't be too hard to add with this approach (compared to #3910)?

bvpvamsikrishna commented:
@michaelcahill @agorrod 
Request to check updates made for cursor operations post conditions as well.
I need to add automated tests, I have tested some operations manually.

agorrod commented on dist/s_string.ok:
```diff
@@ -945,6 +946,7 @@ mkdir
 mmap
 mmrand
 mnt
+mongodb
```
The two changes to this file can be removed.

agorrod commented on src/cursor/cur_log.c:
```diff
@@ -292,8 +292,11 @@ static int
 
 	cl = (WT_CURSOR_LOG *)cursor;
+	session = (WT_SESSION_IMPL *)cursor->session;
+	WT_RET(__wt_txn_context_prepare_check(session, false));
```
I'm surprised the log cursor needs a change for prepare.

agorrod commented on src/include/misc.i:
```diff
@@ -193,15 +193,44 @@ __wt_snprintf_len_incr(
 		    "%s: only permitted in a running transaction",
 		    session->name);
-	if (!requires_txn && F_ISSET(&session->txn, WT_TXN_RUNNING))
+#ifdef HAVE_TIMESTAMPS
```
Why didn't you call `__wt_txn_context_prepare_check` rather than duplicating the code?

bvpvamsikrishna commented:
@agorrod @sueloverso I believe this pull request is ready for review, request to review.
I have added the list of session level API's which can / cannot be invoked in a prepared transaction state in the jira ticket [WT-3848](https://jira.mongodb.org/browse/WT-3848). Request to review the list and provide your opinion.
None of the cursor API's can be invoked in prepared transaction state irrespective of cursor type.  Request to indicate in the review if any of them should be allowed to be invoked in prepared state.

Currently I have added checks in all the session call's to check for the prepared state.
Alternative is to add this check inside SESSION_API_CALL macro so that code changes will be at minimal places as this macro is called by all the session level API's.  If we take this approach, then some session level API's will check for the transaction running state as well. I felt little odd thinking that with in a session API, we check for transaction state twice, once for prepare state and other for running state. Request your thoughts on this.

If this is too much text to read, we can have a small discussion meeting based on your availability

bvpvamsikrishna commented:
@agorrod @sueloverso  I have committed the changes I was referring to above " moving the prepare check to SESSION_API_CALL". Ready for your review.

sueloverso commented on test/suite/test_prepare02.py:
```diff
@@ -0,0 +1,136 @@
+            self.assertRaises(wiredtiger.WiredTigerError,
+                lambda: self.session.timestamp_transaction(
+                    "commit_timestamp=2a"))
+        self.session.log_printf("Wrote 100 records")
```
It isn't clear to me why `log_printf` and `log_flush` are allowed. There are other, equally non-active-transaction methods that we're forbidding such as `transaction_pinned_range` and some other cursor types, etc. These need comments both in the WT code and test code if they stay.

Some session methods are not listed here, such as `transaction_pinned_range` and `log_flush`. (Related to earlier comment about alphabetizing to make it easier to detect missing methods.)

sueloverso commented on test/suite/test_prepare03.py:
```diff
@@ -0,0 +1,223 @@
+            #            lambda: cursor.set_value(self.genvalue(i)))
+            with self.expectedStderrPattern(preparemsg):
+                self.assertRaises(wiredtiger.WiredTigerError,
+                        lambda: cursor.insert())
```
These can be a single line with `assertRaisesWithMessage`.

bvpvamsikrishna commented:
@sueloverso 
I have addressed your comments. Request to check.

In test_prepare02.py, the session level operations invocation is in the same order as used in function assignment within __open_session method. This will be easy to verify any missing functions.

transaction_pinned_range is the only function that is missing in test_prepare02.py. I have created [WT-3916](https://jira.mongodb.org/browse/WT-3916) to address this as the invocation of transaction_pinned_range is complex.

sueloverso commented:
@bvpvamsikrishna I did not get to this today. I will review it again first thing tomorrow.

sueloverso commented on src/session/session_api.c:
```diff
@@ -291,7 +291,13 @@ __session_reconfigure(WT_SESSION *wt_session, const char *config)
-	SESSION_API_CALL(session, reconfigure, config, cfg);
+	/*
+	 * Even though this operation is not allowed in prepared transaction
+	 * state, true is passed indicating the operation is allowed in
```
I suggest removing some of the repetitiveness in this comment:
* Even though this operation is not allowed while the transaction is prepared,
* we pass in true so that the running transaction check below takes precedence
* over the prepared transaction check.

sueloverso commented on src/session/session_api.c:
```diff
@@ -1469,7 +1476,13 @@ __session_begin_transaction(WT_SESSION *wt_session, const char *config)
 	session = (WT_SESSION_IMPL *)wt_session;
-	SESSION_API_CALL(session, begin_transaction, config, cfg);
+	/*
+	 * Even though this operation is not allowed in prepared transaction
```
Same wording here and the other places below.

sueloverso commented on test/suite/test_prepare02.py:
```diff
@@ -0,0 +1,123 @@
+
+        self.session.create("table:mytable", "key_format=S,value_format=S")
+        cursor = self.session.open_cursor("table:mytable", None)
+        # Session operations not permitted after prepare_transaction
```
I think this comment should be expanded because I thought it specifically referred to the begin/prepare lines directly below it. I think it should:
1. Have a blank line after the open_cursor call.
2. Say something like:
 \# 
\# Test the session methods that are forbidden after the transaction is prepared.
 \#

sueloverso commented on test/suite/test_prepare02.py:
```diff
@@ -0,0 +1,123 @@
+        # Below operations are not supported in prepared state. Operations
+        # are listed in the same order as declared with in session
+        # structure and assigned in open_session. Any function missed below
+        # is to be considered as supported in prepare transaction state.
```
I recommend a bit of a rewording:
 \#
 \# The operations listed below are not supported in the prepared state.
 \#
 \# The operations are listed in the same order as they are declared in the session
 \# structure. Any function missing below is allowed in the prepared state.
 \#

sueloverso commented on test/suite/test_prepare02.py:
```diff
@@ -0,0 +1,123 @@
+            lambda:self.session.transaction_sync(), msg)
+        self.session.rollback_transaction()
+
+        # Commit after prepare is permitted
```
Minor nit - add a period at the end of the sentences in the next three comments.

sueloverso commented on test/suite/test_prepare03.py:
```diff
@@ -0,0 +1,211 @@
+            print('**** ERROR in session.create("' + name + '","' + args + '") ***** ')
+            raise
+
+    # Create and session and test cursor operations.
```
I think the comment here should be 'Create the table and test cursor operations.'

sueloverso commented on test/suite/test_prepare03.py:
```diff
@@ -0,0 +1,211 @@
+        self.assertRaisesWithMessage(
+            wiredtiger.WiredTigerError, cursor.get_value, valuemsg)
+
+    def session_create(self, name, args):
```
Was this just for your debugging purposes? It doesn't seem to be needed in this test since the create is not expected to fail. I think it can be removed now.

sueloverso commented on test/suite/test_prepare03.py:
```diff
@@ -0,0 +1,211 @@
+            #cursor.set_value(self.genvalue(i))
+            #with self.expectedStderrPattern(preparemsg):
+            #    self.assertRaises(wiredtiger.WiredTigerError,
+            #            lambda: cursor.set_value(self.genvalue(i)))
```
These commented out lines can be removed.

sueloverso commented on test/suite/test_prepare03.py:
```diff
@@ -0,0 +1,211 @@
+            valformat = 'value_format=8t'
+        else:
+            valformat = 'value_format=S'
+        create_args = keyformat + ',' + valformat
```
This sort of thing should be part of the scenarios so that you can just access them in one location. You can see an example in `test_timestamp04.py` in the `types` section rather than hardcoded here.

sueloverso commented on test/suite/test_prepare03.py:
```diff
@@ -0,0 +1,211 @@
+        self.assertEqual(cursor.uri, tablearg)
+
+        # Check insert and reset operations
+        for i in range(0, self.nentries):
```
While 10 is pretty small is there a reason (I see it is used below for next/prev) to check the error case every iteration? If it works the first insert, it should work on all the later ones too.

sueloverso commented on test/suite/test_prepare03.py:
```diff
@@ -0,0 +1,211 @@
+        # Check search, update, remove, reserve, reconfigure operations.
+        cursor.reset()
+        self.assertCursorHasNoKeyValue(cursor)
+
```
I suggest setting key to `nentries / 2` in case we decide even 10 entries is too many.
A comment here about what you're doing with these keys/values would be helpful.
 \# Search for a specific key.
 \# Verify we get the expected error and then later we can update and remove it.

sueloverso commented on test/suite/test_prepare03.py:
```diff
@@ -0,0 +1,211 @@
+        self.assertRaisesWithMessage(wiredtiger.WiredTigerError,
+            lambda:cursor.search_near(), preparemsg)
+        self.session.commit_transaction()
+        if self.uri == 'lsm':
```
We definitely need a comment here about why LSM needs this handling here.

agorrod commented on src/cursor/cur_std.c:
```diff
@@ -823,6 +823,7 @@ __wt_cursor_close(WT_CURSOR *cursor)
 
 	session = (WT_SESSION_IMPL *)cursor->session;
 
+	WT_RET(__wt_txn_context_prepare_check(session, false));
```
I don't believe WT_CURSOR::close should do a prepare check - we want applications to be able to release resources, and this API call doesn't rely on any prepare semantics.

agorrod commented:
Thanks @bvpvamsikrishna lgtm - I'm going to merge.
